python_modules/dagster/dagster/_config/structured_config/__init__.py

```diff
@@ -783,6 +783,27 @@ def _wrap_config_type(
         raise NotImplementedError(f"Pydantic shape type {shape_type} not supported.")
 
 
+def _get_inner_field_if_exists(
+    shape_type: PydanticShapeType, field: ModelField
+) -> Optional[ModelField]:
+    """Grabs the inner Pydantic field type for a data structure such as a list or dictionary.
+
+    Returns None for types which have no inner field.
+    """
+    # See https://github.com/pydantic/pydantic/blob/v1.10.3/pydantic/fields.py#L758 for
+    # where sub_fields is set.
+    if shape_type == SHAPE_SINGLETON:
+        return None
+    elif shape_type == SHAPE_LIST:
+        # List has a single subfield, which is the type of the list elements.
+        return check.not_none(field.sub_fields)[0]
+    elif shape_type in MAPPING_TYPES:
+        # Mapping has a single subfield, which is the type of the mapping values.
+        return check.not_none(field.sub_fields)[0]
```

Comment:
is the single sub_field typed as as tuple?

Reply:
No, it's typed to whatever the value type is:

```python
>>> class MyModel(BaseModel):
...     foo: Dict[str, int]
...
>>> MyModel.__fields__
{'foo': ModelField(name='foo', type=Mapping[str, int], required=True)}
>>> MyModel.__fields__['foo'].sub_fields
[ModelField(name='_foo', type=int, required=True)]
```

Reply:
Ah how strange